File: CaptionStudio/Services/TimelineSnap.swift
import Foundation

/// Snap helpers for the zoomable timeline (P1.2).
/// Keep algorithm in sync with `enhancer-server/src/timelineSnap.js`.
enum TimelineSnap {
    static let defaultThreshold: TimeInterval = 0.08
    static let pixelSlop: Double = 8
    static let minThreshold: TimeInterval = 0.04

    struct Result: Equatable {
        var start: TimeInterval
        var end: TimeInterval
        var guide: TimeInterval?
        var snapped: Bool
    }

    /// Pixel-aware threshold: at least `minThreshold`, or ~8px in time.
    static func threshold(pixelsPerSecond: Double) -> TimeInterval {
        let pps = max(pixelsPerSecond, 0.001)
        return max(minThreshold, pixelSlop / pps)
    }

    /// Snap moving `[start, end]` to the nearest anchor (start or end edge).
    static func snapInterval(
        start: TimeInterval,
        end: TimeInterval,
        anchors: [TimeInterval],
        threshold: TimeInterval = defaultThreshold
    ) -> Result {
        let length = max(0.05, end - start)
        var s = start
        var bestGuide: TimeInterval?
        var bestViaStart = true
        var bestDist = TimeInterval.infinity

        for anchor in anchors where anchor.isFinite {
            let dStart = abs(s - anchor)
            if dStart < bestDist {
                bestDist = dStart
                bestGuide = anchor
                bestViaStart = true
            }
            let e = s + length
            let dEnd = abs(e - anchor)
            if dEnd < bestDist {
                bestDist = dEnd
                bestGuide = anchor
                bestViaStart = false
            }
        }

        guard let guide = bestGuide, bestDist <= threshold + 1e-9 else {
            return Result(start: s, end: s + length, guide: nil, snapped: false)
        }

        if bestViaStart {
            s = guide
        } else {
            s = guide - length
        }
        return Result(start: s, end: s + length, guide: guide, snapped: true)
    }

    /// Collect edges from sibling clips plus playhead / media bounds.
    static func collectAnchors(
        clips: [(id: String, start: TimeInterval, end: TimeInterval)],
        excluding: Set<String>,
        playhead: TimeInterval?,
        mediaDuration: TimeInterval?
    ) -> [TimeInterval] {
        var anchors: [TimeInterval] = [0]
        if let playhead, playhead.isFinite { anchors.append(playhead) }
        if let mediaDuration, mediaDuration > 0 { anchors.append(mediaDuration) }
        for clip in clips where !excluding.contains(clip.id) {
            anchors.append(clip.start)
            anchors.append(clip.end)
        }
        return anchors
    }
}

File: enhancer-server/src/timelineSnap.js
/**
 * Timeline snap helpers — keep in sync with CaptionStudio/Services/TimelineSnap.swift
 */
export const DEFAULT_SNAP_THRESHOLD_SECONDS = 0.08;

/**
 * Snap a moving interval [start, end] to the nearest anchor.
 * Prefers snapping the leading edge (start); if end is closer, shifts by the same delta.
 * @returns {{ start: number, end: number, guide: number|null, snapped: boolean }}
 */
export function snapInterval(start, end, anchors = [], threshold = DEFAULT_SNAP_THRESHOLD_SECONDS) {
  const length = Math.max(0.05, Number(end) - Number(start));
  let s = Number(start);
  let best = null;
  let bestDist = Infinity;

  for (const raw of anchors) {
    const a = Number(raw);
    if (!Number.isFinite(a)) continue;
    const dStart = Math.abs(s - a);
    if (dStart < bestDist) {
      bestDist = dStart;
      best = { guide: a, via: "start" };
    }
    const e = s + length;
    const dEnd = Math.abs(e - a);
    if (dEnd < bestDist) {
      bestDist = dEnd;
      best = { guide: a, via: "end" };
    }
  }

  if (!best || bestDist > threshold + 1e-9) {
    return { start: s, end: s + length, guide: null, snapped: false };
  }

  if (best.via === "start") {
    s = best.guide;
  } else {
    s = best.guide - length;
  }
  return {
    start: s,
    end: s + length,
    guide: best.guide,
    snapped: true,
  };
}

/** Collect snap anchors from clip edges + playhead + media bounds. */
export function collectSnapAnchors({
  clips = [],
  excludeIds = new Set(),
  playhead = null,
  mediaDuration = null,
  includeZero = true,
} = {}) {
  const anchors = [];
  if (includeZero) anchors.push(0);
  if (Number.isFinite(playhead)) anchors.push(Number(playhead));
  if (Number.isFinite(mediaDuration) && mediaDuration > 0) {
    anchors.push(Number(mediaDuration));
  }
  for (const c of clips) {
    const id = c?.id;
    if (id != null && excludeIds.has(id)) continue;
    if (Number.isFinite(c?.start)) anchors.push(Number(c.start));
    if (Number.isFinite(c?.end)) anchors.push(Number(c.end));
  }
  return anchors;
}

/** Pixel-aware threshold: at least `minSeconds`, or `pixelSlop / pps`. */
export function snapThresholdSeconds(pixelsPerSecond, pixelSlop = 8, minSeconds = 0.04) {
  const pps = Math.max(Number(pixelsPerSecond) || 0, 0.001);
  return Math.max(minSeconds, pixelSlop / pps);
}
